Add E2E tests for component cleanup

tpantelis opened the pull request:
Delete the `Submariner` resource and verify the uninstall `Daemonset` pods successfully complete. If one doesn't, output its status and pod log. Also verify resources are deleted from the broker.

After the test completes, it re-creates the `Submariner` resource.

Added a similar test for `ServiceDiscovery`.

Depends on https://github.com/submariner-io/lighthouse/pull/691

tpantelis committed:
Add E2E test for Submariner component cleanup

Delete the Submariner resource and verify the uninstall Daemonset pods
successfully complete. If one doesn't, output its status and pod log.
Also verify resources are deleted from the broker.

After the test completes, it re-creates the Submariner resource.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

sridhargaddam commented on test/e2e/cleanup/submariner.go:
```diff
@@ -0,0 +1,264 @@
+		brokerRestConfig = getBrokerRestConfig(submariner.Spec.BrokerK8sRemoteNamespace)
+		Expect(brokerRestConfig).ToNot(BeNil(), "No broker located")
+
+		routeAgentPodMonitor = startPodMonitor(names.AppendUninstall(names.RouteAgentComponent), stopCh)
```
Shall we include globalnetPod as well, as it will cover even routeAgent and gatewayPods?

tpantelis replied:
yeah I was going to add globalnet as well.

sridhargaddam commented on test/e2e/cleanup/submariner.go:
```diff
@@ -0,0 +1,264 @@
+		}
+	})
+
+	It("should run DaemonSets/Deployments to uninstall components", func() {
```
Both the pods are DaemonSets.

tpantelis replied:
The network plugin syncer is a `Deployment` although we're not checking that right now..

sridhargaddam commented on test/e2e/cleanup/submariner.go:
```diff
@@ -0,0 +1,264 @@
+
+		_, err = submarinerClient.SubmarinerV1().Clusters(framework.TestContext.SubmarinerNamespace).Get(context.TODO(),
+			clusterID, metav1.GetOptions{})
+		Expect(apierrors.IsNotFound(err)).To(BeTrue(), fmt.Sprintf("Unexpected Cluster %q found on the broker", clusterID))
```
```suggestion
		Expect(apierrors.IsNotFound(err)).To(BeTrue(), fmt.Sprintf("Unexpected Cluster %q found", clusterID))
```

We are validating in the local cluster (not on broker) right?

tpantelis replied:
It's actually meant to check the broker NS. We really should check both to be complete.

sridhargaddam commented on test/e2e/cleanup/submariner.go:
```diff
@@ -0,0 +1,264 @@
+
+		for i := range endpoints.Items {
+			Expect(endpoints.Items[i].Spec.ClusterID).ToNot(Equal(clusterID),
+				fmt.Sprintf("Unexpected Endpoint %q found on the broker", endpoints.Items[i].Name))
```
```suggestion
				fmt.Sprintf("Unexpected Endpoint %q found", endpoints.Items[i].Name))
```

sridhargaddam reviewed (changes requested):
Thanks for the PR @tpantelis. It catches the gateway pod issue too :D

tpantelis committed:
Refactor uninstall logic

...to add a state for each component so they can progress
independently.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Add E2E test for ServiceDiscovery cleanup

Delete the ServiceDiscovery resource and verify the uninstall Deployment
pod successfully completes. If one doesn't, output its status and pod log.
After the test completes, it re-creates the Submariner resource.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

dfarrell07 commented on Makefile:
```diff
@@ -53,9 +53,9 @@ endif
+override E2E_ARGS += $(CLUSTER_SETTINGS_FLAG) cluster1 cluster2
 export DEPLOY_ARGS
-override UNIT_TEST_ARGS += cmd internal/cli internal/env internal/log
+override UNIT_TEST_ARGS += test internal/env
```
These `Makefile` and `e2e.sh` changes are intentional/necessary? They don't seem to be covered by the commit message, or at least it's not clear to me.

tpantelis replied:
They're necessary. Adding "test" to UNIT_TEST_ARGS excludes the E2E tests from being executed by `make unit`. We do this in the other repos. I removed `internal/cli` and `internal/log` from exclusion b/c there are no unit tests in those dirs anyway - I'm unclear why they were added here in first place.

dfarrell07 replied:
That makes sense, thanks @tpantelis.

> I'm unclear why they were added here in first place

That was my main concern, didn't want to lose something there. Thanks for clarifying!

tpantelis commented on scripts/kind-e2e/e2e.sh:
```diff
@@ -69,5 +69,7 @@ ${DAPPER_SOURCE}/bin/subctl benchmark throughput --intra-cluster ${KUBECONFIGS_D
 
 ${DAPPER_SOURCE}/bin/subctl benchmark throughput --verbose ${KUBECONFIGS_DIR}/kind-config-cluster1 ${KUBECONFIGS_DIR}/kind-config-cluster2
 
+${SCRIPTS_DIR}/e2e.sh "$@"
```
@dfarrell07 This is necessary so the new Go E2E tests in _test/e2e_ are executed via `make e2e`.

dfarrell07 replied:
@tpantelis Thanks, that makes a lot of sense. I guess I'm mostly just surprised this didn't exist already.

tpantelis replied:
These are the first Go E2E tests in the repo.

dfarrell07 commented on controllers/uninstall/uninstall.go:
```diff
@@ -38,12 +38,23 @@ const (
+type stateType int
+
+const (
+	deleteComponent stateType = iota
```
Nifty! If anyone else is wondering about `iota`: https://golangbyexample.com/iota-in-golang/

tpantelis replied:
That's the Go way of creating (sort of) enums.

dfarrell07 commented on test/e2e/cleanup/submariner.go:
```diff
@@ -0,0 +1,330 @@
+		pods: map[string]*podInfo{},
+	}
+
+	w, err := watcher.New(&watcher.Config{
```
This pod monitor watcher pattern is interesting, might be a good topic for a code teaching session.

dfarrell07 reviewed (approved):
LGTM
